sdks/go/pkg/beam/io/synthetic/source.go

```diff
@@ -131,8 +130,19 @@ func (fn *sourceFn) ProcessElement(rt *sdf.LockRTracker, config SourceConfig, em
 	for i := rt.GetRestriction().(offsetrange.Restriction).Start; rt.TryClaim(i) == true; i++ {
 		key := make([]byte, config.KeySize)
 		val := make([]byte, config.ValueSize)
-		if _, err := fn.rng.Read(key); err != nil {
-			return err
+		generator := sourceFn{}
+		generator.rng = rand.New(rand.NewSource(i))
```

Comment:
I would avoid making local instances of rand.Rand here unless you have a very good reason here. Is the reason you're creating new, local instances of rand.Rand and seeding them with the restriction positions because you want to get completely deterministic behavior, regardless of any splitting that might occur on the restriction?

If so: Then be aware that this implementation is going to get identical results per input (in this case, per SourceConfig). For example, if you make two SourceConfigs for 10 elements each, then both of them will seed these generators with the numbers 0-9, which can lead to strangely identical results. If you want purely deterministic output that is still different per SourceConfig, then each SourceConfig will probably need some kind of unique but deterministic salt that can be added to these numbers (example: `rand.NewSource(i + fn.salt))`).

On the other hand, if this check and the hotkey distributions don't need to be deterministic, then I would simplify a bit and just use the existing Rand object (`fn.rng`) more. Or, if you need to customize the rng for test purposes, you can add more `randWrapper` fields to the SourceFn struct. I think the only spot where you would need deterministic values in that case is for filling out the actual hotkey value, and even then I think there's room for optimizing so we don't constantly create and reseed a rand.Rand.

Reply:
@youngoli thank you for your feedback. Since I have the most experience with Python, when I was implementing it I was following the way it was implemented in Python.

In Python generator is seed with each index, therefore I did the same in Go, and I seed the generator with a NewSource on each iteration. There is no salt in config in other SDKs implementation and the idea behind it was: To be consistent with other SDKs to be able to compare them.
```python
  def _gen_kv_pair(self, generator, index):
    generator.seed(index)
    rand = generator.random_sample()

    # Determines whether to generate hot key or not.
    if rand < self._hot_key_fraction:
      # Generate hot key.
      # An integer is randomly selected from the range [0, numHotKeys-1]
      # with equal probability.
      generator_hot = Generator(index % self._num_hot_keys)
      bytes_ = generator_hot.bytes(self._key_size), generator.bytes(
        self._value_size)
    else:
      bytes_ = generator.bytes(self.element_size)
      bytes_ = bytes_[:self._key_size], bytes_[self._key_size:]
    return bytes_

  def read(self, range_tracker):
    index = range_tracker.start_position()
    generator = Generator()
    while range_tracker.try_claim(index):
      time.sleep(self._sleep_per_input_record_sec)
      yield self._gen_kv_pair(generator, index)
      index += 1
```

Reply:
Hm... I have objections with the way that one is coded too, but that's something for me to discuss with the original author. It's definitely reasonable to match the Python implementation.

In that case, my only remaining advice is, instead of creating a new Rand object with a new source each time you want to reseed (in other words, in each position), create the Rand objects in the beginning of `ProcessElement` and then call [Rand.Seed](https://golang.org/pkg/math/rand/#Rand.Seed) when you want to reseed.

Reply:
@youngoli  Thank you, I pushed the changes you suggested.